R/print.data.table.R

```diff
@@ -205,6 +172,55 @@ paste_dims = function(x) {
   paste0("[", paste(dims,collapse="x"), "]")
 }
 
+format_col = function(x, ...) {
+  UseMethod("format_col")
+}
+
+format_list_item = function(x, ...) {
+  UseMethod("format_list_item")
+}
+
+format_col.default = function(x, ...) {
+  if (!is.null(dim(x))) return("<multi-column>")
+  if (is.list(x)) return(vapply_1c(x, format_list_item, ...))
+  format(char.trunc(x), ...) # relevant to #37
+}
+
+# #2842 -- different columns can have different tzone, so force usage in output
+format_col.POSIXct = function(x, ..., timezone=FALSE) {
+  if (timezone) {
+    tz = attr(x,'tzone',exact=TRUE)
+    nas = is.na(x)
+    x = paste0(as.character(x)," ",tz)
+    x[nas] = NA_character_
+  } else {
+    x = format(x, usetz=FALSE)
+  }
+  x
+}
+
+# #3011 -- expression columns can wrap to newlines which breaks printing
+format_col.expression = function(x, ...) format(char.trunc(as.character(x)), ...)
+
+format_list_item.default = function(x, ...) {
+  if (is.null(x))  # NULL item in a list column
+    ""
+  else if (is.atomic(x) || inherits(x, "formula")) # FR #2591 - format.data.table issue with columns of class "formula"
+    paste(c(format(head(x, 6L), ...), if (length(x) > 6L) "..."), collapse=",")  # fix for #5435 and #37 - format has to be added here...
+  else
+    paste0("<", class(x)[1L], paste_dims(x), ">")
+}
+
+# FR #1091 for pretty printing of character
+# TODO: maybe instead of doing "this is...", we could do "this ... test"?
+char.trunc <- function(x, trunc.char = getOption("datatable.prettyprint.char")) {
+  trunc.char = max(0L, suppressWarnings(as.integer(trunc.char[1L])), na.rm=TRUE)
+  if (!is.character(x) || trunc.char <= 0L) return(x)
+  idx = which(nchar(x) > trunc.char)
```

Comment:
not crucial for this PR in particular, but wouldn't `nchar(x, 'width')` be more appropriate?

Reply:
Good idea. Would need a test too in a new PR: your Chinese multi-byte data?